drivers: serial: added UART interrupts for Raspberry Pi Pico board

andrei-edward-popa opened the pull request:
Added UART Interrupt Driven driver for Raspberry Pi Pico board.

Signed-off-by: Andrei-Edward Popa <andrei_edward.popa@upb.ro>

gmarull commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -14,62 +15,313 @@
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
 };
 
+#define DEV_CFG(dev)  \
+	((const struct uart_rpi_config *const)(dev)->config)
+#define DEV_DATA(dev) \
+	((struct uart_rpi_data *const)(dev)->data)
+#define DEV_UART_INST(dev) \
+	((uart_inst_t *const)((const struct uart_rpi_config *const)(dev)->config)->uart_dev)
+#define DEV_UART_HW(dev) \
+	uart_get_hw(DEV_UART_INST(dev))
```
please remove this sort of accessors, ref. https://github.com/zephyrproject-rtos/zephyr/pull/41918

andrei-edward-popa replied:
Ok, I removed all accessors.

gmarull commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -14,62 +15,313 @@
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),	\
+		.irq_config_func = uart##idx##_rpi_irq_config_func	\
+	}
+
+#define RPI_UART_CONFIG_FUNC(idx)	\
+	static void uart##idx##_rpi_irq_config_func(const struct device *port)	\
+	{	\
+		IRQ_CONNECT(DT_INST_IRQN(idx),	\
+			    DT_INST_IRQ(idx, priority),	\
+			    uart_rpi_isr,	\
+			    DEVICE_DT_INST_GET(idx), 0);	\
+		irq_enable(DT_INST_IRQN(idx));	\
+	}
+
+#else
+
+#define RPI_UART_CREATE_CFG(idx)	\
+	static const struct uart_rpi_config uart##idx##_rpi_config = {	\
+		.uart_dev = uart##idx,	\
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),	\
+	}
+
+#define RPI_UART_CONFIG_FUNC(idx)
+
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+#define RPI_UART_INIT(idx)	\
+	PINCTRL_DT_INST_DEFINE(idx);	\
+	static struct uart_rpi_data uart##idx##_rpi_data = {	\
+		.baudrate = DT_INST_PROP(idx, current_speed),	\
+	};	\
+	\
+	RPI_UART_CONFIG_FUNC(idx)	\
+	RPI_UART_CREATE_CFG(idx);	\
+	\
+	DEVICE_DT_INST_DEFINE(idx, &uart_rpi_init,	\
+			    NULL, &uart##idx##_rpi_data,	\
+			    &uart##idx##_rpi_config, PRE_KERNEL_1,	\
+			    CONFIG_SERIAL_INIT_PRIORITY,	\
+			    &uart_rpi_driver_api);	\
```
please align `\` to the same column for readability

andrei-edward-popa replied:
I aligned it before, but the pipeline check said that I have a line with more than 100 characters, but it was only 77 characters long. I will try to align it in a different way.

andrei-edward-popa replied:
I found that the tab width was the problem. Now should be okay, you can check it.

gmarull commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -14,62 +15,313 @@
+	static const struct uart_rpi_config uart##idx##_rpi_config = {	\
+		.uart_dev = uart##idx,	\
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),	\
+		.irq_config_func = uart##idx##_rpi_irq_config_func	\
+	}
+
+#define RPI_UART_CONFIG_FUNC(idx)	\
+	static void uart##idx##_rpi_irq_config_func(const struct device *port)	\
+	{	\
+		IRQ_CONNECT(DT_INST_IRQN(idx),	\
+			    DT_INST_IRQ(idx, priority),	\
+			    uart_rpi_isr,	\
+			    DEVICE_DT_INST_GET(idx), 0);	\
+		irq_enable(DT_INST_IRQN(idx));	\
+	}
+
+#else
+
+#define RPI_UART_CREATE_CFG(idx)	\
+	static const struct uart_rpi_config uart##idx##_rpi_config = {	\
+		.uart_dev = uart##idx,	\
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),	\
```
this is duplicating initializer fields unnecessarily, better use something like:

```
#ifdef CONFIG_UART_INTERRUPT_DRIVEN
#define RPI_UART_IRQ_CONFIG_INIT .irq_config_func = ...,
#else
#define RPI_UART_IRQ_CONFIG_INIT
#endif
```

or, make use of `IF_ENABLED(...)`

andrei-edward-popa replied:
Done!

gmarull commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -14,62 +15,307 @@
+	uart_hw_t * const uart_hw = uart_get_hw(uart_inst);
 
-	if (!uart_is_readable(config->uart_dev)) {
+	if (!uart_is_readable(uart_inst)) {
```
why is this change needed? it doesn't seem related to interrupt support.

andrei-edward-popa replied:
This change is not needed, sorry for that.

gmarull commented on boards/arm/rpi_pico/rpi_pico.dts:
```diff
@@ -14,6 +14,7 @@
 		zephyr,sram = &sram0;
 		zephyr,flash = &flash0;
 		zephyr,console = &uart0;
+		zephyr,shell-uart = &uart0;
```
please add board changes to another commit

andrei-edward-popa replied:
Okay.

andrei-edward-popa commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -14,62 +15,307 @@
+	int tx_len = 0;
+
+	const struct uart_rpi_config * const config = dev->config;
+	uart_inst_t * const uart_inst = config->uart_dev;
+	uart_hw_t * const uart_hw = uart_get_hw(uart_inst);
+
+	while (uart_is_writable(uart_inst) && (len - tx_len) > 0) {
+		uart_hw->dr = tx_data[tx_len++];
+	}
```
Yes, uart_hw_t is the register structure that can be accessed by uart_get_hw API function. I will call this function anytime I need, I will remove uart_hw pointer from everywhere.

gmarull replied:
my initial concern has not been addressed: with these changes, the driver is mixing high-level APIs with register access, which, I think, is confusing. If high-level API is not suitable for Zephyr drivers, then everything should be moved to register based for consistency.

andrei-edward-popa replied:
I followed the approach from the polling method right now.

andrei-edward-popa commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -14,62 +15,307 @@
 	}
 
-	uart_init(config->uart_dev, config->baudrate);
+	baudrate = uart_init(uart_inst, data->baudrate);
+	/* Check if baudrate adjustment returned by 'uart_init' function is a positive value */
+	if (baudrate <= 0) {
+		return -EINVAL;
+	}
```
I added a separate commit for this modification.

gmarull commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -70,7 +74,7 @@ static int uart_rpi_init(const struct device *dev)
 #endif /* CONFIG_UART_INTERRUPT_DRIVEN */
 
-	return 0;
+	return ret;
```
this can stay to `0`

andrei-edward-popa replied:
Okay, I will let it 0.

gmarull commented on drivers/serial/Kconfig.rpi_pico:
```diff
@@ -9,3 +9,4 @@ config UART_RPI_PICO
 	default $(dt_compat_enabled,$(DT_COMPAT_RPI_PICO_UART))
 	select SERIAL_HAS_DRIVER
 	select PICOSDK_USE_UART
+	select SERIAL_SUPPORT_INTERRUPT
```
this is up to the board to decide, I guess

andrei-edward-popa replied:
In all drivers that I saw in zephyr this was here.

gmarull replied:
sorry, got confused with `CONFIG_UART_INTERRUPT_DRIVEN`, my fault.

gmarull commented on boards/arm/rpi_pico/rpi_pico_defconfig:
```diff
@@ -7,3 +7,4 @@ CONFIG_UART_CONSOLE=y
 CONFIG_GPIO=y
 CONFIG_USE_DT_CODE_PARTITION=y
 CONFIG_BUILD_OUTPUT_UF2=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
```
this is updating board: should go into a separate commit

andrei-edward-popa replied:
Can be in the same commit with dts modifications?

andrei-edward-popa replied:
Done!

gmarull replied:
nit: for bisectability this commit needs to go after adding interrupt support

andrei-edward-popa replied:
Okay, makes sense.

gmarull commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -39,37 +50,249 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
+{
+	const struct uart_rpi_config * const config = dev->config;
+
+	uart_get_hw(config->uart_dev)->imsc |= (UART_UARTIMSC_OEIM_BITS |
+											UART_UARTIMSC_BEIM_BITS |
+											UART_UARTIMSC_PEIM_BITS |
+											UART_UARTIMSC_FEIM_BITS |
+											UART_UARTIMSC_RTIM_BITS);
```
looks like formatting needs to be improved here

andrei-edward-popa replied:
Yes, I formatted it.

Commit:
boards: arm: rpi_pico: added zephyr shell

added zephyr shell to be used on uart0 instance

Signed-off-by: Andrei-Edward Popa <andrei_edward.popa@upb.ro>

Commit:
boards: arm: rpi_pico: added UART interrupt driven config

enabled UART interrupt driven

Signed-off-by: Andrei-Edward Popa <andrei_edward.popa@upb.ro>

Commit:
drivers: serial: rpi_pico: check if baudrate was set by the API function

if peripheral clock is not configured, uart init API function returns 0,
so we need to check the return value of this function

Signed-off-by: Andrei-Edward Popa <andrei_edward.popa@upb.ro>

Commit:
drivers: serial: added UART interrupts for Raspberry Pi Pico board

added UART interrupts driver for Raspberry Pi Pico board
moved baudrate from config to data structure

Signed-off-by: Andrei-Edward Popa <andrei_edward.popa@upb.ro>

pirearadu commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -53,27 +66,238 @@ static int uart_rpi_init(const struct device *dev)
+static int uart_rpi_fifo_fill(const struct device *dev,
+			      const uint8_t *tx_data, int len)
+{
+	int tx_len = 0;
```
Please use reverse christmas tree declaration.

andrei-edward-popa replied:
Okay, I will modify everywhere.

pirearadu commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -53,27 +66,238 @@ static int uart_rpi_init(const struct device *dev)
+{
+	int rx_len = 0;
+
+	const struct uart_rpi_config * const config = dev->config;
```
Please use reverse christmas tree declaration.

andrei-edward-popa replied:
Done!

pirearadu commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -53,27 +66,238 @@ static int uart_rpi_init(const struct device *dev)
+{
+	int errors = 0;
+
+	const struct uart_rpi_config * const config = dev->config;
```
Please use reverse christmas tree declaration.
And no space between declarations.

andrei-edward-popa replied:
Done!

gmarull commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -40,36 +51,249 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
+	uart_inst_t * const uart_inst = config->uart_dev;
+	int tx_len = 0;
+
+	while (uart_is_writable(uart_inst) && (len - tx_len) > 0) {
```
this driver continues to mix API and register access. I think it would be better to just use register access everywhere, so for example:

```suggestion
	while (!(uart_get_hw(uart)->fr & UART_UARTFR_TXFF_BITS) && (len - tx_len) > 0) {
```

`uart_get_hw` could actually be removed by just storing a `uart_hw_t *`.

andrei-edward-popa replied:
Okay, I understand now, I will do this modifications soon.

andrei-edward-popa replied:
@gmarull What about uart_init function?

gmarull replied:
note: `uart_init()` actions should also be handled by the driver

andrei-edward-popa replied:
Note that uart_init function is using reset and clock API functions, but reset driver and clock driver are not implemented in Zephyr yet. If I want to implement the init function I need to bring another API functions. I think for now this API call is not dangerous and in the future I would like to implement those drivers in Zephyr and then I will modify also this function.

gmarull replied:
Fair enough, I guess it's ok for now then.

gmarull replied:
Would you mind adding a comment around `uart_init()`?

andrei-edward-popa replied:
A comment like?

gmarull replied:
```
/* uart_init() may be replaced by register based API once rpi-pico platform has a clock controller driver */
```

andrei-edward-popa replied:
Done! Thanks for review. :) Tomorrow I will make a PR with I2C driver.

Commit:
drivers: serial: rpi_pico: replaced high level API functions

replaced high level API functions with register access

Signed-off-by: Andrei-Edward Popa <andrei_edward.popa@upb.ro>

andrei-edward-popa commented:
@dcpleung Hi! Can you review my code? :)

dcpleung commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -40,6 +52,9 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
 static int uart_rpi_init(const struct device *dev)
 {
 	const struct uart_rpi_config *config = dev->config;
+	uart_inst_t * const uart_inst = config->uart_dev;
```
To keep the tree bisect-able, this should be in the previous commit where the call to `uart_init()` was added.

andrei-edward-popa replied:
Yes, you are right, it was my mistake. It is okay now.

dcpleung commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -40,6 +52,9 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
 	const struct uart_rpi_config *config = dev->config;
+	uart_inst_t * const uart_inst = config->uart_dev;
+	uart_hw_t * const uart_hw = config->uart_regs;
+	struct uart_rpi_data * const data = dev->data;
```
This also needs to be in the previous commit.

andrei-edward-popa replied:
Done!

dcpleung reviewed (changes requested):
LGTM but please move the introduction of variables to the correct commit.
